handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -612,108 +509,162 @@ public final SSLEngineResult wrap(
-                // present.
+                // Explicit use outboundClosed as we want to drain any bytes that are still present.
                 if (outboundClosed) {
-                    return drainOutboundBuffer(dst, status);
+                    // There is something left to drain.
+                    // See https://github.com/netty/netty/issues/6260
+                    bytesProduced = SSL.bioFlushByteBuffer(networkBIO);
+                    return newResultMayFinishHandshake(NOT_HANDSHAKING, 0, bytesProduced);
                 }
-            }
 
-            int endOffset = offset + length;
-            int srcsLen = 0;
+                // Flush any data that may be implicitly generated by OpenSSL (handshake, close, etc..).
+                SSLEngineResult.HandshakeStatus status = NOT_HANDSHAKING;
+                // Prepare OpenSSL to work in server mode and receive handshake
+                if (handshakeState != HandshakeState.FINISHED) {
+                    if (handshakeState != HandshakeState.STARTED_EXPLICITLY) {
+                        // Update accepted so we know we triggered the handshake via wrap
+                        handshakeState = HandshakeState.STARTED_IMPLICITLY;
+                    }
 
-            for (int i = offset; i < endOffset; ++i) {
-                final ByteBuffer src = srcs[i];
-                if (src == null) {
-                    throw new IllegalArgumentException("srcs[" + i + "] is null");
-                }
-                if (srcsLen == MAX_PLAINTEXT_LENGTH) {
-                    continue;
+                    // Flush any data that may have been written implicitly during the handshake by OpenSSL.
+                    bytesProduced = SSL.bioFlushByteBuffer(networkBIO);
+
+                    // Check to see if the engine wrote data into the network BIO.
+                    if (bytesProduced > 0) {
+                        // We produced / consumed some data during the handshake, signal back to the caller.
+                        // If there is a handshake exception and we have produced data, we should send the data before
+                        // we allow handshake() to throw the handshake exception.
+                        if (handshakeException == null) {
+                            status = handshake();
+
+                            if (renegotiationPending && status == FINISHED) {
+                                // If renegotiationPending is true that means when we attempted to start renegotiation
+                                // the BIO buffer didn't have enough space to hold the HelloRequest which prompts the
+                                // client to initiate a renegotiation. At this point the HelloRequest has been written
+                                // so we can actually start the handshake process.
+                                renegotiationPending = false;
+                                SSL.setState(ssl, SSL.SSL_ST_ACCEPT);
```

Comment:
The OpenSSL team has advised against doing this. I have fixed the current approach in this PR and will investigate removal in of `SSL.SSL_ST_ACCEPT` in https://github.com/netty/netty/issues/6320.